### ISSUE TITLE
Deleting a database crashes server

### ISSUE THREAD

**Issue by miyconst** [Starcounter/Starcounter#3522]
I tried to delete a database called `warehouse`, the database was not started. The following exception occurs and administrator page hangs. After administrator page reload the is no more databases listed.

After a couple more reloads Starcounter server crashed.

I had another database called `default` running at this time, only SignIn app inside, if this helps anyhow.

### Deleting database log entires

```
"Starcounter.DbException: ScErrDatabaseRunning (SCERR10031): The database is running. Database: 'sc://konstantin-pc/personal/warehouse'.
Version: 2.2.1852.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR10031.
   at Starcounter.Server.Commands.DeleteDatabaseCommandProcessor.CheckPrerequisites(Database& database, DeleteDatabaseCommand& command) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Server\Starcounter.Server\Commands\Processors\DeleteDatabaseCommandProcessor.cs:line 78
   at Starcounter.Server.Commands.DeleteDatabaseCommandProcessor.Execute() in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Server\Starcounter.Server\Commands\Processors\DeleteDatabaseCommandProcessor.cs:line 27
   at Starcounter.Server.Commands.CommandProcessor.ProcessCommand(NotifyCommandStatusChangedCallback notifyStatusChangedCallback) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Server\Starcounter.Server\Commands\CommandProcessor.cs:line 300
ErrorCode=10031
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR10031
"
```

```
"ScErrServerCommandFailed (SCERR10011): The operation ""Deleting database warehouse (including data files)"" failed.
Version: 2.2.1852.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR10011."
```

### Server crash log entires

```
Gateway XML: MaximumReceiveContentLength parameter exceeds maximum value. Current value is set to 2096936.
```

```
ScErrUnhandledException (SCERR1069): An unhandled exception was detected.
```

```
ScErrDatabaseEngineTerminated (SCERR10007): The server detected that the user code host executable (sccode.exe) unexpectedly terminated during a management operation.
```

```
"System.Exception: ScErrUnhandledException (SCERR1069): An unhandled exception was detected.
Version: 2.2.1852.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR1069.
   at StarcounterInternal.Hosting.Processor.RunTaskNoScheduling(Action userAction, String appName) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Hosting\Starcounter.Hosting\Processor.cs:line 102
   at Starcounter.DbSession.GetAndExecuteQueuedTasks(Action`2 taskExecutionMethod) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter\DbSession.cs:line 180
   at StarcounterInternal.Hosting.Processor.RunTask(IntPtr hTask) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Hosting\Starcounter.Hosting\Processor.cs:line 149
   at StarcounterInternal.Hosting.Processor.RunMessageLoop(Void* hsched) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Hosting\Starcounter.Hosting\Processor.cs:line 62
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR1069
---> System.Exception: ScErrAcquireSessionTimeout (SCERR5008): Exclusive access to the session could not be obtained.
Version: 2.2.1852.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR5008.
   at Starcounter.Session.StartUsing() in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.XSON\Sessions\Session.cs:line 528
   at Starcounter.Session.Use(SessionTask action, String sessionId) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.XSON\Sessions\Session.cs:line 710
   at Starcounter.Task.Run() in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter\DbSession.cs:line 100
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR5008
"
```

```
"System.Exception: ScErrAcquireSessionTimeout (SCERR5008): Exclusive access to the session could not be obtained.
Version: 2.2.1852.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR5008.
   at Starcounter.Session.StartUsing() in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.XSON\Sessions\Session.cs:line 528
   at Starcounter.Session.ScheduleTask(String sessionId, SessionTask task, Boolean waitForCompletion) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.XSON\Sessions\Session.cs:line 319
   at Starcounter.Session.ScheduleTask(IEnumerable`1 sessionIds, SessionTask task, Boolean waitForCompletion) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.XSON\Sessions\Session.cs:line 336
   at Administrator.Server.Model.Server.OnChanged(Object sender, EventArgs e) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.Administrator\Server\Model\Server.cs:line 196
   at Administrator.Server.Model.Database.OnChanged(Object sender, EventArgs e) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.Administrator\Server\Model\Database.cs:line 393
   at Administrator.Server.Model.Database.Database_PropertyChanged(Object sender, PropertyChangedEventArgs e) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.Administrator\Server\Model\Database.cs:line 378
   at Administrator.Server.Model.Database.OnPropertyChanged(String fieldName) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.Administrator\Server\Model\Database.cs:line 1247
   at Administrator.Server.Model.Database.ExecuteCommand(ServerCommand command, Action`1 completionCallback, Action`5 errorCallback) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.Administrator\Server\Model\Database.cs:line 1157
   at Starcounter.Templates.Property`1.InvokeHandler(Json parent, Input`1 input) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.XSON\Templates\Foundation\Property.cs:line 242
   at Starcounter.Templates.Property`1.ProcessInput(Json parent, T value) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.XSON\Templates\Foundation\Property.cs:line 264
   at Starcounter.XSON.DefaultPatchHandler.ParseAndProcess(TLong property, Json parent, IntPtr valuePtr, Int32 valueSize) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.XSON\Sessions\JSON-Patch\DefaultPatchHandler.cs:line 105
   at Starcounter.Advanced.XSON.JsonExtension.Scope(Json json, Action action) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.XSON\Class Json\JsonExtension.cs:line 36
   at Starcounter.XSON.DefaultPatchHandler.Handle(Json root, JsonPatchOperation patchOp, JsonPointer pointer, IntPtr valuePtr, Int32 valueSize) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.XSON\Sessions\JSON-Patch\DefaultPatchHandler.cs:line 37
   at Starcounter.XSON.JsonPatch.Apply(Json root, IntPtr patchArrayPtr, Int32 patchArraySize, Boolean strictPatchRejection) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.XSON\Sessions\JSON-Patch\JsonPatch.cs:line 650
   at Starcounter.XSON.JsonPatch.Apply(Json root, Byte[] patchArray, Boolean strictPatchRejection) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.XSON\Sessions\JSON-Patch\JsonPatch.cs:line 491
   at Administrator.Server.Managers.ServerManager.<>c.<RegisterFullModelApi>b__7_2(String data, WebSocket ws) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.Administrator\Server\Managers\ServerManager.cs:line 143
   at Starcounter.Internal.AllWsGroups.RunHandler(UInt16 managedHandlerId, UInt32 groupId, WebSocket ws) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.Internal\Http\WsManager.cs:line 289
   at Starcounter.GatewayHandlers.HandleWebSocket(UInt16 managedHandlerId, Byte* rawChunk, BMX_TASK_INFO* taskInfo, Boolean* isHandled) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter\Apps\GatewayHandlers.cs:line 784
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR5008
"
```

```
"System.Exception: ScErrUnhandledException (SCERR1069): An unhandled exception was detected.
Version: 2.2.1852.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR1069.
   at StarcounterInternal.Hosting.Processor.RunTaskNoScheduling(Action userAction, String appName) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Hosting\Starcounter.Hosting\Processor.cs:line 102
   at Starcounter.DbSession.GetAndExecuteQueuedTasks(Action`2 taskExecutionMethod) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter\DbSession.cs:line 180
   at StarcounterInternal.Hosting.Processor.RunTask(IntPtr hTask) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Hosting\Starcounter.Hosting\Processor.cs:line 149
   at StarcounterInternal.Hosting.Processor.RunMessageLoop(Void* hsched) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Hosting\Starcounter.Hosting\Processor.cs:line 62
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR1069
---> System.Exception: ScErrAcquireSessionTimeout (SCERR5008): Exclusive access to the session could not be obtained.
Version: 2.2.1852.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR5008.
   at Starcounter.Session.StartUsing() in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.XSON\Sessions\Session.cs:line 528
   at Starcounter.Session.Use(SessionTask action, String sessionId) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.XSON\Sessions\Session.cs:line 710
   at Starcounter.Task.Run() in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter\DbSession.cs:line 100
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR5008
"
```

cc @chrhol, @andwah, @alemoi

**Comment by miyconst** [comment]
After server restart I was able to delete both databases, but deleting `default` database again rose an exception and required page reload.

```
"System.Exception: ScErrAcquireSessionTimeout (SCERR5008): Exclusive access to the session could not be obtained.
Version: 2.2.1852.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR5008.
   at Starcounter.Session.StartUsing() in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.XSON\Sessions\Session.cs:line 528
   at Starcounter.Session.Use(SessionTask action, String sessionId) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.XSON\Sessions\Session.cs:line 710
   at Starcounter.Session.ScheduleTask(String sessionId, SessionTask task, Boolean waitForCompletion) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.XSON\Sessions\Session.cs:line 287
   at Starcounter.Session.ScheduleTask(IEnumerable`1 sessionIds, SessionTask task, Boolean waitForCompletion) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.XSON\Sessions\Session.cs:line 336
   at Administrator.Server.Model.Server.OnChanged(Object sender, EventArgs e) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.Administrator\Server\Model\Server.cs:line 196
   at System.Collections.ObjectModel.ObservableCollection`1.OnCollectionChanged(NotifyCollectionChangedEventArgs e)
   at System.Collections.ObjectModel.ObservableCollection`1.RemoveItem(Int32 index)
   at System.Collections.ObjectModel.Collection`1.Remove(T item)
   at Administrator.Server.Model.Server.InvalidateDatabases() in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.Administrator\Server\Model\Server.cs:line 244
   at Administrator.Server.Managers.ServerManager.<>c.<RegisterInternalModelApi>b__9_1(String databaseName, Request request) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.Administrator\Server\Managers\ServerManager.cs:line 294
   at lambda_method(Closure , Request , IntPtr , IntPtr )
   at Starcounter.Rest.UserHandlerInfo.RunUserDelegate(Request req, IntPtr methodSpaceUriSpaceOnStack, IntPtr parametersInfoOnStack) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.Rest\UriHandlersManager.cs:line 221
   at Starcounter.Internal.Web.AppRestServer.RunDelegateAndProcessResponse(IntPtr methodSpaceUriSpaceOnStack, IntPtr parametersInfoOnStack, Request req) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.Apps.JsonPatch\AppRestServer.cs:line 98
   at Starcounter.Internal.AppsBootstrapper.ProcessExternalRequest(Request req) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-285\Level1\src\Starcounter.Apps.JsonPatch\AppsBootstrapper.cs:line 598
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR5008
"
```

**Comment by andwah** [comment]
It's looks it has something todo with pushing out async changes to the client(s).  There has been recent changes on how things is scheduled and how sessions is used.

**Comment by chrhol** [comment]
Looks like something is not releasing the exclusive lock to a session. I will take a look

**Comment by chrhol** [comment]
The problem is a deadlock in acquiring access to a session and the locks in `ServerManager.cs`. I found one issue with the current scheduling and that is sometimes a task is scheduled with the same session specified as the current one. 

As soon as my local build succeeds I will push a fix for scheduling the same session. I have not been able to reproduce it again after this so hopefully this is the culprit.

I used another bug to be able to reproduce this issue though, that the administrator only looks if `sccode.exe` is running to decide if a database is running or not, and not both `sccode.exe` and `scdata.exe`. I will add another issue.

**Comment by chrhol** [comment]
Pushed to develop, soon pnext

**Comment by andwah** [comment]
> I used another bug to be able to reproduce this issue though, that the administrator only looks if sccode.exe is running to decide if a database is running or not, and not both sccode.exe and scdata.exe. I will add another issue.

In #2913 it's explained how we check if a database is running or not.

Admin's check looks like this;
```
bool isRunning = (databaseInfo != null && databaseInfo.Engine != null && databaseInfo.Engine.DatabaseProcessRunning && databaseInfo.Engine.HostProcessId != 0);
```

**Comment by andwah** [comment]
Just got this errors , this is ***after*** i pulled the latest code changes in (develop) by @chrhol 

```
20160421T142513 Critical sc://headsutv19.innowait.se/personal/administrator Starcounter - System.Exception: ScErrUnhandledException (SCERR1069): An unhandled exception was detected.
Version: 2.0.0.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR1069.
   at StarcounterInternal.Hosting.Processor.RunTaskNoScheduling(Action userAction, String appName) in C:\\github\\Level1\\src\\Hosting\\Starcounter.Hosting\\Processor.cs:line 120
   at Starcounter.DbSession.GetAndExecuteQueuedTasks(Action`2 taskExecutionMethod) in C:\\github\\Level1\\src\\Starcounter\\DbSession.cs:line 180
   at StarcounterInternal.Hosting.Processor.RunTask(IntPtr hTask) in C:\\github\\Level1\\src\\Hosting\\Starcounter.Hosting\\Processor.cs:line 149
   at StarcounterInternal.Hosting.Processor.RunMessageLoop(Void* hsched) in C:\\github\\Level1\\src\\Hosting\\Starcounter.Hosting\\Processor.cs:line 62
   HResult=-2146233088
   HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR1069
   ---> System.Exception: ScErrAcquireSessionTimeout (SCERR5008): Exclusive access to the session could not be obtained.
   Version: 2.0.0.
   Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR5008.
   at Starcounter.Session.StartUsing() in C:\\github\\Level1\\src\\Starcounter.XSON\\Sessions\\Session.cs:line 531
   at Starcounter.Session.Use(SessionTask action, String sessionId) in C:\\github\\Level1\\src\\Starcounter.XSON\\Sessions\\Session.cs:line 715
   at Starcounter.Session.<>c__DisplayClass25_0.<ScheduleTask>b__0() in C:\\github\\Level1\\src\\Starcounter.XSON\\Sessions\\Session.cs:line 310
   at Starcounter.Task.Run() in C:\\github\\Level1\\src\\Starcounter\\DbSession.cs:line 100
   HResult=-2146233088
   HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR5008
   
20160421T142513 Critical sc://headsutv19.innowait.se/personal/administrator Starcounter - System.Exception: ScErrUnhandledException (SCERR1069): An unhandled exception was detected.
Version: 2.0.0.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR1069.
   at StarcounterInternal.Hosting.Processor.RunTaskNoScheduling(Action userAction, String appName) in C:\\github\\Level1\\src\\Hosting\\Starcounter.Hosting\\Processor.cs:line 120
   at Starcounter.DbSession.GetAndExecuteQueuedTasks(Action`2 taskExecutionMethod) in C:\\github\\Level1\\src\\Starcounter\\DbSession.cs:line 180
   at StarcounterInternal.Hosting.Processor.RunTask(IntPtr hTask) in C:\\github\\Level1\\src\\Hosting\\Starcounter.Hosting\\Processor.cs:line 149
   at StarcounterInternal.Hosting.Processor.RunMessageLoop(Void* hsched) in C:\\github\\Level1\\src\\Hosting\\Starcounter.Hosting\\Processor.cs:line 62
   HResult=-2146233088
   HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR1069
   ---> System.Exception: ScErrAcquireSessionTimeout (SCERR5008): Exclusive access to the session could not be obtained.
   Version: 2.0.0.
   Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR5008.
   at Starcounter.Session.StartUsing() in C:\\github\\Level1\\src\\Starcounter.XSON\\Sessions\\Session.cs:line 531
   at Starcounter.Session.Use(SessionTask action, String sessionId) in C:\\github\\Level1\\src\\Starcounter.XSON\\Sessions\\Session.cs:line 715
   at Starcounter.Session.<>c__DisplayClass25_0.<ScheduleTask>b__0() in C:\\github\\Level1\\src\\Starcounter.XSON\\Sessions\\Session.cs:line 310
   at Starcounter.Task.Run() in C:\\github\\Level1\\src\\Starcounter\\DbSession.cs:line 100
   HResult=-2146233088
   HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR5008
   
20160421T142513 Critical sc://headsutv19.innowait.se/personal/administrator Starcounter - System.Exception: ScErrUnhandledException (SCERR1069): An unhandled exception was detected.
Version: 2.0.0.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR1069.
   at StarcounterInternal.Hosting.Processor.RunTaskNoScheduling(Action userAction, String appName) in C:\\github\\Level1\\src\\Hosting\\Starcounter.Hosting\\Processor.cs:line 120
   at Starcounter.DbSession.GetAndExecuteQueuedTasks(Action`2 taskExecutionMethod) in C:\\github\\Level1\\src\\Starcounter\\DbSession.cs:line 180
   at StarcounterInternal.Hosting.Processor.RunTask(IntPtr hTask) in C:\\github\\Level1\\src\\Hosting\\Starcounter.Hosting\\Processor.cs:line 149
   at StarcounterInternal.Hosting.Processor.RunMessageLoop(Void* hsched) in C:\\github\\Level1\\src\\Hosting\\Starcounter.Hosting\\Processor.cs:line 62
   HResult=-2146233088
   HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR1069
   ---> System.Exception: ScErrAcquireSessionTimeout (SCERR5008): Exclusive access to the session could not be obtained.
   Version: 2.0.0.
   Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR5008.
   at Starcounter.Session.StartUsing() in C:\\github\\Level1\\src\\Starcounter.XSON\\Sessions\\Session.cs:line 531
   at Starcounter.Session.Use(SessionTask action, String sessionId) in C:\\github\\Level1\\src\\Starcounter.XSON\\Sessions\\Session.cs:line 715
   at Starcounter.Session.<>c__DisplayClass25_0.<ScheduleTask>b__0() in C:\\github\\Level1\\src\\Starcounter.XSON\\Sessions\\Session.cs:line 310
   at Starcounter.Task.Run() in C:\\github\\Level1\\src\\Starcounter\\DbSession.cs:line 100
   HResult=-2146233088
   HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR5008
   
20160421T142513 Error sc://headsutv19/personal Starcounter.Host 5008 System.Exception: ScErrAcquireSessionTimeout (SCERR5008): Exclusive access to the session could not be obtained.
Version: 2.0.0.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR5008.
   at Starcounter.Session.StartUsing() in C:\\github\\Level1\\src\\Starcounter.XSON\\Sessions\\Session.cs:line 531
   at Starcounter.Session.Use(SessionTask action, String sessionId) in C:\\github\\Level1\\src\\Starcounter.XSON\\Sessions\\Session.cs:line 715
   at Starcounter.Session.ScheduleTask(String sessionId, SessionTask task, Boolean waitForCompletion) in C:\\github\\Level1\\src\\Starcounter.XSON\\Sessions\\Session.cs:line 292
   at Starcounter.Session.ScheduleTask(IEnumerable`1 sessionIds, SessionTask task, Boolean waitForCompletion) in C:\\github\\Level1\\src\\Starcounter.XSON\\Sessions\\Session.cs:line 343
   at Administrator.Server.Managers.ServerManager.ServerModel_Changed(Object sender, EventArgs e) in C:\\github\\Level1\\src\\Starcounter.Administrator\\Server\\Managers\\ServerManager.cs:line 331
   at Administrator.Server.Model.Server.OnChanged(Object sender, EventArgs e) in C:\\github\\Level1\\src\\Starcounter.Administrator\\Server\\Model\\Server.cs:line 194
   at Administrator.Server.Model.Server.Databases_CollectionChanged(Object sender, NotifyCollectionChangedEventArgs e) in C:\\github\\Level1\\src\\Starcounter.Administrator\\Server\\Model\\Server.cs:line 172
   at System.Collections.ObjectModel.ObservableCollection`1.OnCollectionChanged(NotifyCollectionChangedEventArgs e)
   at System.Collections.ObjectModel.ObservableCollection`1.RemoveItem(Int32 index)
   at System.Collections.ObjectModel.Collection`1.Remove(T item)
   at Administrator.Server.Model.Server.InvalidateDatabases() in C:\\github\\Level1\\src\\Starcounter.Administrator\\Server\\Model\\Server.cs:line 245
   at Administrator.Server.Managers.ServerManager.<>c.<RegisterInternalModelApi>b__9_1(String databaseName, Request request) in C:\\github\\Level1\\src\\Starcounter.Administrator\\Server\\Managers\\ServerManager.cs:line 294
   at lambda_method(Closure , Request , IntPtr , IntPtr )
   at Starcounter.Rest.UserHandlerInfo.RunUserDelegate(Request req, IntPtr methodSpaceUriSpaceOnStack, IntPtr parametersInfoOnStack) in C:\\github\\Level1\\src\\Starcounter.Rest\\UriHandlersManager.cs:line 221
   at Starcounter.Internal.Web.AppRestServer.RunDelegateAndProcessResponse(IntPtr methodSpaceUriSpaceOnStack, IntPtr parametersInfoOnStack, Request req) in C:\\github\\Level1\\src\\Starcounter.Apps.JsonPatch\\AppRestServer.cs:line 98
   at Starcounter.Internal.AppsBootstrapper.ProcessExternalRequest(Request req) in C:\\github\\Level1\\src\\Starcounter.Apps.JsonPatch\\AppsBootstrapper.cs:line 598
   HResult=-2146233088
   HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR5008
 
   
   
20160421T142513 Error sc://headsutv19/personal/service Starcounter - ScErrDatabaseEngineTerminated (SCERR10007): The server detected that the user code host executable (sccode.exe) unexpectedly terminated during a management operation.
20160421T142513 Error sc://headsutv19/personal/service Starcounter - ScErrUnhandledException (SCERR1069): An unhandled exception was detected.
```

**Comment by chrhol** [comment]
Then the deadlock is still not resolved. I wonder if the problem is that we skip scheduling sometimes and execute the task directly if running on the same scheduler.

What did you do to reproduce it this time?

**Comment by andwah** [comment]
i think i found a clue,, there is a timeout in a rest call just before the error, i will investigate some more..

**Comment by chrhol** [comment]
I think we finally found the cause of the deadlock. It looks like it's the same problem as in #3490, that if `Session.ScheduleTask` is called for a session that runs on the same scheduler as the current one it executes the task directly. This together with the locks in parts of the administrator (`ServerManager.cs`) will make it quite easy to end up in a deadlock.

I tested with making sure that no tasks are executed directly and after that I could not reproduce this, and actually all operations feels like they execute a lot faster.

We need to fix the API in `Session` properly. I will add a new issue for that I think, so we can decide on the proper way of doing it.
